compiler/src/dotty/tools/dotc/core/CheckRealizable.scala

```diff
@@ -66,22 +66,53 @@ class CheckRealizable(implicit ctx: Context) {
    */
   private def isLateInitialized(sym: Symbol) = sym.is(LateInitialized, butNot = Module)
 
-  /** The realizability status of given type `tp`*/
+  /** The realizability status of given type `tp`. Types can only project
+    * members from realizable types, that is, types having non-null values and
+    * not depending on mutable state.
+    * Beware that subtypes of realizable types might not be realizable; hence
+    * realizability checking restricts overriding. */
   def realizability(tp: Type): Realizability = tp.dealias match {
     case tp: TermRef =>
+      // Suppose tp is a.b.c.type, where c is declared with type T, then sym is c, tp.info is T and
+      // and tp.prefix is a.b.
       val sym = tp.symbol
-      if (sym.is(Stable)) realizability(tp.prefix)
-      else {
-        val r =
-          if (!sym.isStable) NotStable
-          else if (!isLateInitialized(sym)) Realizable
-          else if (!sym.isEffectivelyFinal) new NotFinal(sym)
-          else realizability(tp.info).mapError(r => new ProblemInUnderlying(tp.info, r))
-        r andAlso {
-          sym.setFlag(Stable)
-          realizability(tp.prefix)
-        }
+      // tp is realizable if it selects a stable member on a realizable prefix.
+
+      /** A member is always stable if tp.info is a realizable singleton type. We check this last
+          for performance, in all cases where some unrelated check might have failed. */
+      def patchRealizability(r: Realizability) =
+        r.mapError(if (tp.info.isStableRealizable) Realizable else _)
```

Comment:
@milessabin You asked whether `def foo[L <: Lambda with Singleton]: L#Arg = ???` could work. That warrants an issue but here's a pointer, while I have this code fresh in memory. In this method we have code that considers realizable many things that are singletons of stable paths. The `SingletonType` branch below has similar logic.

The implementation would be different tho, since we'd have to handle type variables.

And as mentioned, this is blocked on `Singleton` not doing its job because of union types.